### ISSUE TITLE
Bring delta support to OneDrive/Sharepoint remote

### ISSUE THREAD

**Issue by jcastilloalonso** [rclone/rclone#7317]
#### The associated forum post URL from `https://forum.rclone.org`

https://forum.rclone.org/t/possible-listr-support-for-onedrive-sharepoint/41832/4?u=jcastill

#### What is your current rclone version (output from `rclone version`)?

1.64.0

#### What problem are you are trying to solve?

Currently OneDrive does not support ListR, yet it seems that implementing delta (https://learn.microsoft.com/en-us/onedrive/developer/rest-api/concepts/scan-guidance?view=odsp-graph-online) would allow for ListR to be functional on the remote. This would help a lot specially as Microsoft is very conservative on TPSLimit before throttling so a list command of a large folder structure can easily trigger a throttle event.

Seems delta also supports single files/sub folders so it does not need to always scan the whole drive (https://learn.microsoft.com/en-us/graph/api/driveitem-delta?view=graph-rest-1.0&tabs=http)

#### How do you think rclone should be changed to solve that?

This can be used when --fast-list is invoked to scan the whole subfolder in less API calls in lieu of using more RAM. This can also be used for polling once a scan is completed, not sure if it would only work if initial scan is used by delta, in which case it might make sense to make it default for the remote as it's part of Microsoft API best practices.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
This is a great idea - ListR improves the experience massively on tps limited backends.

If we were to implement this for onedrive then calling `vfs/refresh` will fill the cache up using ListR (now that #7318 is (already) done!) and the polling mechanism already part of onedrive will keep it topped up which will pretty much bring it into line with microsoft best practices.

I had a go at a proof of concept for this.

Unfortunately the API doesn't work on anything except the root so it is inefficient to use `onedrive:dir`.

Can you give this a go? If you could time these two commands that would be very useful. For my onedrive it was actually quicker without the ListR but I don't have very many files.

    rclone lsf -R onedrive: > 1
    rclone lsf -R --disable ListR onedrive: > 0

Under unix I'd do this

    time rclone lsf -R onedrive: | sort > 1
    time rclone lsf -R --disable ListR onedrive: | sort > 0

If all wen't well `0` and `1` should have the same contents. In my tests I found that

- some folders are duplicated
- shared folders are missing

Use `-vv --dump bodies` to see what it is doing.

[v1.65.0-beta.7379.bfd439849.fix-7317-onedrive-listr](https://beta.rclone.org/branch/fix-7317-onedrive-listr/v1.65.0-beta.7379.bfd439849.fix-7317-onedrive-listr/) on branch [fix-7317-onedrive-listr](https://github.com/rclone/rclone/tree/fix-7317-onedrive-listr) (uploaded in 15-30 mins)

**Comment by jcastilloalonso** [comment]
Did a quick test, I stopped 2nd command after it was clear time difference was significant:

# time ./rclone lsf -R OneDrive1: | sort > 1

real    6m32.227s
user    1m52.446s
sys     0m3.788s

# time ./rclone lsf -R --disable ListR OneDrive1: | sort > 0

real    23m46.052s
user    0m48.653s
sys     0m8.144s

I'm going to test another (smaller) OneDrive I got.

**Comment by jcastilloalonso** [comment]
It really takes around the same time on the smaller OneDrive:
```
# time ./rclone lsf -R OneD: --log-level DEBUG --log-file ListR.log --dump-bodies > ListR.txt
real    1m48.017s
user    0m9.352s
sys     0m1.794s
# time ./rclone lsf -R --disable ListR OneD: --log-level DEBUG --log-file NoListR.log --dump-bodies > NoListR.txt
real    1m22.970s
user    0m21.867s
sys     0m4.114s
```

Yet debug logs are half the size for new implementation:
```
# ls -lah
-rw-r-----.  1 root     root     113M Sep 19 10:34 ListR.log
-rw-r--r--.  1 root     root     8.5M Sep 19 10:34 ListR.txt
-rw-r-----.  1 root     root     214M Sep 19 10:39 NoListR.log
-rw-r--r--.  1 root     root     8.3M Sep 19 10:39 NoListR.txt
```

Upon sorting the folders, I do also see lot of duplicate folders in the new method. I do not have any shared files in my OneDrive so can't talk about those and if they are missing or not. I'm going to try and search if there's any documentation about this behavior and report back. This can also be a OneDrive API error.

**Comment by ncw** [comment]
Your tests look encouraging.

I fixed the duplicates and the shared items and made it pass the integration tests, so I think it is probably ready now.

**Comment by ncw** [comment]
Pressed enter too soon!

Here is the new test binary - can you give it a go? Also if you could see if it works with vfs/refresh too that would be good.

[v1.65.0-beta.7380.c8d5b159f.fix-7317-onedrive-listr](https://beta.rclone.org/branch/fix-7317-onedrive-listr/v1.65.0-beta.7380.c8d5b159f.fix-7317-onedrive-listr/) on branch [fix-7317-onedrive-listr](https://github.com/rclone/rclone/tree/fix-7317-onedrive-listr) (uploaded in 15-30 mins)

I have a slight reservation about merging this in that ListR only works at the root of the onedrive. If you use it not at the root then it recurses from the root and tosses all the data that is not under the directory you asked for. So it will be correct, I'm just a bit concerned about efficiency.

**Comment by jcastilloalonso** [comment]
Thanks Nick, I'll check the binaries and play around. I'll do some time comparison loading subfolders. Will report back most likely tomorrow.

**Comment by jcastilloalonso** [comment]
I think something is not working correctly.

Mounted with RC and ran the following command:

```
# rclone rc vfs/refresh recursive=true --rc-addr 127.0.0.1:5573 _async=true
{
        "jobid": 1
}
```

After a couple of minutes I checked the status of the job:
```
# rclone rc job/status --json '{ "jobid":1 }' --rc-addr 127.0.0.1:5573
{
        "duration": 0,
        "endTime": "0001-01-01T00:00:00Z",
        "error": "",
        "finished": false,
        "group": "job/1",
        "id": 1,
        "output": null,
        "startTime": "2023-09-20T16:45:23.63038475-05:00",
        "success": false
}

```

I'm logging with debug flag just to see what happened but don't see much activity going on:

```
2023/09/20 16:44:35 DEBUG : Encrypted drive '1USB2:': Mounting on "/mnt/tmp"
2023/09/20 16:44:35 DEBUG : : Root:
2023/09/20 16:44:35 DEBUG : : >Root: node=/, err=<nil>
2023/09/20 16:44:36 DEBUG : /: Attr:
2023/09/20 16:44:36 DEBUG : /: >Attr: attr=valid=1s ino=0 size=0 mode=drwxr-xr-x, err=<nil>
2023/09/20 16:44:36 DEBUG : : Statfs:
2023/09/20 16:44:36 DEBUG : : >Statfs: stat={Blocks:6553337856 Bfree:1087363410 Bavail:1087363410 Files:1000000000 Ffree:1000000000 Bsize:4096 Namelen:255 Frsize:4096}, err=<nil>
2023/09/20 16:44:50 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:45:05 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:45:20 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:45:23 DEBUG : rc: "vfs/refresh": with parameters map[_async:true recursive:true]
2023/09/20 16:45:23 DEBUG : rc: "vfs/refresh": reply map[jobid:1]: <nil>
2023/09/20 16:45:23 DEBUG : : Reading directory tree
2023/09/20 16:45:35 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:45:35 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/09/20 16:45:50 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:46:05 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:46:06 DEBUG : rc: "job/status": with parameters map[jobid:1]
2023/09/20 16:46:06 DEBUG : rc: "job/status": reply map[duration:0 endTime:0001-01-01T00:00:00Z error: finished:false group:job/1 id:1 output:<nil> startTime:2023-09-20T16:45:23.63038475-05:00 success:false]: <nil>
2023/09/20 16:46:20 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:46:35 DEBUG : OneDrive root 'USB2': Checking for changes on remote
2023/09/20 16:46:35 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
```

**Comment by jcastilloalonso** [comment]
There might be a previous issue with vfs/refresh command and OneDrive. As I'm trying with 1.64 and 1.63.1 and getting the same behavior. Somehow it's not triggering the warmup:

```
# rclone rc vfs/refresh recursive=true --rc-addr 127.0.0.1:5573 _async=true
{
        "jobid": 1
}
# rclone rc job/status --json '{ "jobid":1 }' --rc-addr 127.0.0.1:5573
{
        "duration": 0,
        "endTime": "0001-01-01T00:00:00Z",
        "error": "",
        "finished": false,
        "group": "job/1",
        "id": 1,
        "output": null,
        "startTime": "2023-09-21T00:46:15.183062728-05:00",
        "success": false
}

```

```
2023/09/21 00:46:04 INFO  : Starting transaction limiter: max 3 transactions/s with burst 1
2023/09/21 00:46:04 DEBUG : rclone: Version "v1.63.1" starting with parameters ["rclone" "mount" "UIA:" "/mnt/tmp" "--allow-other" "--cache-dir" "/mnt/External/cache2/" "--dir-cache-time" "5000h" "--disable-http2" "--log-file" "/mnt/External/1USB2beta.txt" "--log-level" "DEBUG" "--onedrive-no-versions" "--poll-interval" "15s" "--rc" "--rc-addr" "127.0.0.1:5573" "--uid" "1000" "--use-mmap" "--user-agent" "ISV|rclone|rclone/v1.63.1" "--tpslimit" "3" "--tpslimit-burst" "0" "--vfs-cache-max-age" "9999h" "--vfs-cache-max-size" "2500G" "--vfs-cache-mode" "full" "--vfs-fast-fingerprint" "--vfs-write-back" "1h"]
2023/09/21 00:46:04 NOTICE: Serving remote control on http://127.0.0.1:5573/
2023/09/21 00:46:04 DEBUG : Creating backend with remote "UIA:"
2023/09/21 00:46:04 DEBUG : Using config file from "/root/.config/rclone/rclone.conf"
2023/09/21 00:46:04 DEBUG : UIA: detected overridden config - adding "{kWaC5}" suffix to name
2023/09/21 00:46:04 DEBUG : fs cache: renaming cache item "UIA:" to be canonical "UIA{kWaC5}:"
2023/09/21 00:46:04 DEBUG : OneDrive root '': Next delta token is: xxx
2023/09/21 00:46:04 DEBUG : vfs cache: root is "/mnt/External/cache2"
2023/09/21 00:46:04 DEBUG : vfs cache: data root is "/mnt/External/cache2/vfs/UIA{kWaC5}"
2023/09/21 00:46:04 DEBUG : vfs cache: metadata root is "/mnt/External/cache2/vfsMeta/UIA{kWaC5}"
2023/09/21 00:46:04 DEBUG : Creating backend with remote "/mnt/External/cache2/vfs/UIA{kWaC5}/"
2023/09/21 00:46:04 DEBUG : fs cache: renaming cache item "/mnt/External/cache2/vfs/UIA{kWaC5}/" to be canonical "/mnt/External/cache2/vfs/UIA{kWaC5}"
2023/09/21 00:46:04 DEBUG : Creating backend with remote "/mnt/External/cache2/vfsMeta/UIA{kWaC5}/"
2023/09/21 00:46:04 DEBUG : fs cache: renaming cache item "/mnt/External/cache2/vfsMeta/UIA{kWaC5}/" to be canonical "/mnt/External/cache2/vfsMeta/UIA{kWaC5}"
2023/09/21 00:46:04 INFO  : vfs cache: cleaned: objects 0 (was 0) in use 0, to upload 0, uploading 0, total size 0 (was 0)
2023/09/21 00:46:04 DEBUG : OneDrive root '': Mounting on "/mnt/tmp"
2023/09/21 00:46:04 DEBUG : : Root:
2023/09/21 00:46:04 DEBUG : : >Root: node=/, err=<nil>


2023/09/21 00:46:15 DEBUG : rc: "vfs/refresh": with parameters map[_async:true recursive:true]
2023/09/21 00:46:15 DEBUG : rc: "vfs/refresh": reply map[jobid:1]: <nil>
2023/09/21 00:46:15 DEBUG : : Reading directory tree
2023/09/21 00:46:19 DEBUG : OneDrive root '': Checking for changes on remote
2023/09/21 00:46:26 DEBUG : rc: "job/status": with parameters map[jobid:1]
2023/09/21 00:46:26 DEBUG : rc: "job/status": reply map[duration:0 endTime:0001-01-01T00:00:00Z error: finished:false group:job/1 id:1 output:<nil> startTime:2023-09-21T00:46:15.183062728-05:00 success:false]: <nil>
2023/09/21 00:46:34 DEBUG : OneDrive root '': Checking for changes on remote
2023/09/21 00:46:49 DEBUG : OneDrive root '': Checking for changes on remote
2023/09/21 00:47:04 DEBUG : OneDrive root '': Checking for changes on remote

```

**Comment by jcastilloalonso** [comment]
Never mind, tested again with more patience and seems to be working:

```
# rclone rc vfs/refresh recursive=true --rc-addr 127.0.0.1:5573 _async=true
{
        "jobid": 1
}
# rclone rc job/status --json '{ "jobid":1 }' --rc-addr 127.0.0.1:5573
{
        "duration": 96.656083187,
        "endTime": "2023-09-21T01:00:35.116286286-05:00",
        "error": "",
        "finished": true,
        "group": "job/1",
        "id": 1,
        "output": {
                "result": {
                        "": "OK"
                }
        },
        "startTime": "2023-09-21T00:58:58.460203037-05:00",
        "success": true
}
```
From Logs:

`2023/09/21 01:00:35 DEBUG : : Reading directory tree done in 1m36.655803435s`

I will retry with my bigger mounts and report again.

**Comment by jcastilloalonso** [comment]
OMG This is an amazing difference, I got rid of the TPSLimit and TPSLimit-Burst to test, and see how fast it might get throttled, but it never did.

`2023/09/21 01:11:09 DEBUG : : Reading directory tree done in 5m44.957874479s`

As as sidenote on previous code it would take close to 1 hour to do a whole vfs/refresh of this mount with multiple throttling events. 5 minutes is just an amazing difference.

Was going to try the old method again to compare, but after 2 minutes got the following:

```
2023/09/21 01:16:34 DEBUG : >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
2023/09/21 01:16:34 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2023/09/21 01:16:34 DEBUG : HTTP RESPONSE (req 0xc002167500)
2023/09/21 01:16:34 DEBUG : HTTP/1.1 429
Transfer-Encoding: chunked
Cache-Control: private
Client-Request-Id: xxx
Content-Type: application/json
Date: Thu, 21 Sep 2023 06:16:34 GMT
Request-Id: xxx
Retry-After: 15
Strict-Transport-Security: max-age=31536000
Vary: Accept-Encoding
X-Ms-Ags-Diagnostic: {"ServerInfo":{"DataCenter":"South Central US","Slice":"E","Ring":"5","ScaleUnit":"002","RoleInstance":"xxxx"}}

127
{"error":{"code":"activityLimitReached","message":"The request has been throttled","innerError":{"code":"throttledRequest","innerError":{"code":"quota"},"date":"2023-09-21T06:16:34","request-id":"xxxx","client-request-id":"xxxx"}}}

```

Using delta the whole warmup went without a single throttle event. I imagine as Microsoft expects the whole file structure to be read. This is a huge boost

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.65

Thank you all for testing :-)

**Comment by ncw** [comment]
Due to #7362 I have decided that this feature needs to be optionally enabled.

I have added a `--onedrive-delta` flag as suggested which will default to off. So you will need to enable this feature with `delta = true` in the config file or `--onedrive-delta` on the command line.

This will go into 1.65
